lib/fog/openstack/network.rb

```diff
@@ -310,18 +323,29 @@ def request(params)
         def authenticate
           if !@openstack_management_url || @openstack_must_reauthenticate
             options = {
-              :openstack_tenant   => @openstack_tenant,
-              :openstack_api_key  => @openstack_api_key,
-              :openstack_username => @openstack_username,
-              :openstack_auth_uri => @openstack_auth_uri,
-              :openstack_auth_token => @openstack_must_reauthenticate ? nil : @openstack_auth_token,
-              :openstack_service_type => @openstack_service_type,
-              :openstack_service_name => @openstack_service_name,
+              :openstack_tenant        => @openstack_tenant,
+              :openstack_tenant_id      => @openstack_tenant_id,
+              :openstack_api_key       => @openstack_api_key,
+              :openstack_username      => @openstack_username,
+              :openstack_user_domain    => @openstack_user_domain,
+              :openstack_prj_domain     => @openstack_prj_domain,
+              :openstack_auth_uri      => @openstack_auth_uri,
+              :openstack_auth_token    => @openstack_must_reauthenticate ? nil : @openstack_auth_token,
+              :openstack_service_type  => @openstack_service_type,
+              :openstack_service_name  => @openstack_service_name,
               :openstack_endpoint_type => @openstack_endpoint_type,
-              :openstack_region => @openstack_region
+              :openstack_region        => @openstack_region
             }
 
-            credentials = Fog::OpenStack.authenticate_v2(options, @connection_options)
+            case @openstack_auth_uri.path
+            when %r{/v2.0}
+              credentials = Fog::OpenStack.authenticate_v2(options, @connection_options)
+            when %r{/v3}
+              credentials = Fog::OpenStack.authenticate_v3(options, @connection_options)
+            else
+              credentials = Fog::OpenStack.authenticate_v2(options, @connection_options)
```

Comment:
yeah. I'm not sure which one to select either if the path do not give the information. the thing is v3 means require a domain, while v2 do not expect this.  less constraint was my decision. but I may be wrong.